docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,233 @@
+
+This is not to say the P/Invoke system should be completely redesigned. The current system is heavily used and its simplicity for consuming native assets is a benefit. Rather this new mechanism is designed to provide a way for marshaling code to be generated by an external tool but integrate with `DllImportAttribute` in a way that isn't onerous on current .NET developers.
+
+The [Roslyn Compiler](https://github.com/dotnet/roslyn) team is working on a [Source Generator feature][source_gen_link] that will allow the generation of additional source files that can be added to an assembly during the compilation process - the runtime generation IL Stubs is an in-memory version of this scenario.
+
+**Note** This proposal is targeted at addressing P/Invoke improvements but could be adapted to work with COM interop utilizing the new [`ComWrappers`][comwrappers_link] API.
+
+### Requirements
+
+* [Source generators][source_gen_link]
+  * Branch: https://github.com/dotnet/roslyn/tree/features/source-generators
+
+## Design
+
+Using Source Generators is focused on integrating with existing uses of `DllImportAttribute` from an invocation point of view (i.e. callsites should not need to be updated). The idea behind Source Generators is that code for some scenarios can be precomputed using user declared types and logic thus avoiding the need to generator code at runtime. The desire is then to provide a way that existing code can continue to function but can be modified in a way that allows it to leverage this new compiler feature.
+
+### P/Invoke Walkthrough
+
+The P/Invoke algorithm is presented below using a simple example.
+
+``` CSharp
+/* A */ [DllImportAttribute("Kernel32.dll")]
+/* B */ extern static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* C */ QueryPerformanceCounter(out count);
+```
+
+At (A) in the above code snippet, the runtime is told to look for an export name `QueryPerformanceCounter` (B) in the `Kernel32.dll` binary. There are many additional attributes on the `DllImportAttribute` that help with export discovery and can influence the semantics of the generated IL Stub. Point (C) represents an invocation of the P/Invoke. Most of the work occurs at (C) at runtime, since (A) and (B) are merely declarations the compiler uses to embed the relevant details into assembly metadata that is read at runtime.
+
+1) During invocation, the function declaration is determined to be an external call requiring marshaling. Given the defined properties in the `DllImportAttribute` instance as well as the metadata of the user-defined signature an IL Stub is generated.
+
+    * An IL Stub is always generated when an assembly is compiled in `Debug`. In `Release` builds it is possible no IL Stub is generated and instead the JIT will elide the stub and inline the invocation. We will come back to this inlining below.
```

Comment:
This goes into CoreCLR implementation details. Not sure whether this helps the clarity. Skipping the IL stub generation is a historic shortcut that is not really required.

Conceptually, the IL stub is always generated. In Release builds, the JIT inlines the IL stub like any other method if it is profitable and possible.